pkg/kubectl/cmd/apply/prune/prune.go

```diff
@@ -0,0 +1,230 @@
+		if len(gvk) != 3 {
+			return nil, fmt.Errorf("invalid GroupVersionKind format: %v, please follow <group/version/kind>", groupVersionKind)
+		}
+		if gvk[0] == "core" {
+			gvk[0] = ""
+		}
+		parsedGvks = append(parsedGvks, gvk)
+	}
+	return parsedGvks, nil
+}
+
+// Prune resources not previously visited.
+func (p *Pruner) Prune(visitedNamespaces sets.String, visitedUids sets.String) error {
+	// Prune namespaced resources
+	for n := range visitedNamespaces {
+		for _, m := range p.GroupVersionKinds {
+			if m.Scope.Name() == meta.RESTScopeNameNamespace {
+				if err := p.pruneMapping(n, m, visitedUids); err != nil {
+					return fmt.Errorf("error pruning namespaced object %v: %v", m.GroupVersionKind, err)
+				}
+			}
+		}
+	}
+	// Prune non-namespaced resources
+	for _, m := range p.GroupVersionKinds {
+		if m.Scope.Name() == meta.RESTScopeNameRoot {
+			if err := p.pruneMapping(metav1.NamespaceNone, m, visitedUids); err != nil {
+				return fmt.Errorf("error pruning non-namespaced object %v: %v", m.GroupVersionKind, err)
+			}
+		}
+	}
+	return nil
+}
+
+// Deletes objects in namespace/group/version/kind which have not previously been visited.
+func (p *Pruner) pruneMapping(namespace string, mapping *meta.RESTMapping, visitedUids sets.String) error {
+	c, err := p.ClientFunc(mapping)
+	if err != nil {
+		return err
+	}
+
+	export := false
+	apiVersion := mapping.GroupVersionKind.Version
+	objList, err := resource.NewHelper(c, mapping).List(
+		namespace,
+		apiVersion,
+		export,
+		&metav1.ListOptions{
+			LabelSelector:        p.Selector,
+			IncludeUninitialized: p.IncludeUninitialized,
+		},
+	)
+	if err != nil {
+		return err
+	}
+	objs, err := meta.ExtractList(objList)
+	if err != nil {
+		return err
+	}
+
+	for _, obj := range objs {
+		annots, err := mapping.MetadataAccessor.Annotations(obj)
+		if err != nil {
+			return err
+		}
+		if _, ok := annots[api.LastAppliedConfigAnnotation]; !ok {
+			// don't prune resources not created with apply
+			// TODO: log resource not pruned because not created with apply
+			continue
+		}
+		uid, err := mapping.UID(obj)
+		if err != nil {
+			return err
+		}
+		if visitedUids.Has(string(uid)) {
+			// TODO: log resource not pruned because already visited.
+			continue
+		}
+
+		name, err := mapping.Name(obj)
+		if err != nil {
+			return err
+		}
+		if !p.DryRun {
+			if err := p.delete(namespace, name, mapping); err != nil {
+				return err
+			}
+		}
+		cmdutil.PrintSuccess(p.Mapper, p.ShortOutput, p.Out, mapping.Resource, name, p.DryRun, "pruned")
+	}
+	return nil
+}
+
+func (p *Pruner) delete(namespace, name string, mapping *meta.RESTMapping) error {
+
+	c, err := p.ClientFunc(mapping)
+	if err != nil {
+		return err
+	}
+
+	helper := resource.NewHelper(c, mapping)
+	if !p.Cascade {
+		return helper.Delete(namespace, name)
+	}
+
+	cs, err := p.ClientsetFunc()
+	if err != nil {
+		return err
+	}
+	r, err := kubectl.ReaperFor(mapping.GroupVersionKind.GroupKind(), cs)
```

Comment:
Check we @caesarxuchao if we even still need this.  Garbage collection should exist for most resources now.

Reply:
Try the foreground deleteOptions: https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion

Reply:
Same question asked by Clayton: https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!topic/kubernetes-sig-cli/f3eyKD0TrUU
I guess we can remove the reaper code in 1.10.

Reply:
Let's keep the exact same code, and not change the functionality in this PR, in order to make the PR low risk. Once merged I'll open an issue and address it in a future PR.